### PR TITLE
Refactor object data and metadata get to StorageLayer

### PR DESCRIPTION
**Pull request by tonyjhuang**
### Description

Push Rules auth responsibility into `StorageLayer` for GET object data/metadata requests.

The overarching proposal is to move all business logic into `StorageLayer`. The handlers in `firebase.ts` will be responsible for:

- parsing incoming request objects and passing them to the appropriate `StorageLayer`
- translating the request objects form `StorageLayer` into `express` `res` objects

### CODE REVIEW

File: src/emulator/storage/files.ts [hunk 1]
```diff
@@ -162,6 +182,37 @@ export class StorageLayer {
...
+    if (!authorized) {
+      authorized = await isPermitted({
+        ruleset: this._rules,
+        method: RulesetOperationMethod.GET,
+        path: operationPath,
+        file: { before: metadata?.asRulesResource() },
+        authorization: request.authorization,
+      });
+    }
+    if (!authorized) {
```
Comment (tohhsinpei) [inline src/emulator/storage/files.ts:204]: I think it's worth leaving in the comments about auth via header vs. token for web vs. mobile

File: src/emulator/storage/apis/firebase.ts [hunk 2]
```diff
@@ -126,86 +88,52 @@ export function createFirebaseEndpoints(emulator: StorageEmulator): Router {
...
+        bucketId: req.params.bucketId,
+        authorization: req.header("authorization"),
+        downloadToken: req.query.token?.toString(),
+      }));
+    } catch (err) {
+      if (err instanceof NotFoundError) {
+        return res.sendStatus(404);
+      } else if (err instanceof ForbiddenError) {
+        return res.sendStatus(403);
+      } else {
```
Comment (tohhsinpei) [inline src/emulator/storage/apis/firebase.ts:106]: No need `else`

**Review by tohhsinpei — COMMENTED** [review]
Mostly fine, left some nits

File: src/emulator/storage/apis/firebase.ts [hunk 3]
```diff
@@ -126,86 +88,52 @@ export function createFirebaseEndpoints(emulator: StorageEmulator): Router {
...
-
       if (rangeStart) {
         const range = {
           start: parseInt(rangeStart),
           end: rangeEnd ? parseInt(rangeEnd) : data.byteLength,
         };
         res.setHeader("Content-Range", `bytes ${range.start}-${range.end - 1}/${data.byteLength}`);
-        res.status(206).end(data.slice(range.start, range.end));
+        return res.status(206).end(data.slice(range.start, range.end));
       } else {
```
[the remark this answers is not in the document]
Comment (tonyjhuang) [reply src/emulator/storage/apis/firebase.ts:130]: done

File: src/emulator/storage/apis/firebase.ts (hunk 2 above)
Comment (tonyjhuang) [reply src/emulator/storage/apis/firebase.ts:106]: done

**Review by tohhsinpei — APPROVED** [review]

**Pull request merged**